### PR TITLE
012 add new role

### PR DESCRIPTION
**Pull request by Ryan-Jones1999**

### CODE REVIEW

File: src/main/java/com/kainos/ea/controller/JobController.java [hunk 1]
```diff
@@ -3,27 +3,32 @@
 import com.kainos.ea.dao.JobDao;
 import com.kainos.ea.exception.DatabaseException;
 import com.kainos.ea.exception.NotAValidBandLevelException;
+import com.kainos.ea.exception.RoleNotAddedException;
+import com.kainos.ea.model.JobRole;
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/controller/JobController.java:7]: remove unused import

File: src/main/java/com/kainos/ea/controller/JobController.java [hunk 2]
```diff
@@ -3,27 +3,32 @@
...
-import javax.ws.rs.GET;
-import javax.ws.rs.Path;
-import javax.ws.rs.PathParam;
-import javax.ws.rs.Produces;
+import javax.ws.rs.*;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
 import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/controller/JobController.java:19]: Remove unused imports lines 18 -19

File: src/main/java/com/kainos/ea/controller/JobController.java [hunk 3]
```diff
@@ -82,4 +87,61 @@ public Response viewCompentency (@PathParam("bandlevelID") int bandID) {
...
+    @POST
+    @Path("/addnewrole")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response addNewRole (NewRoleRequest newrolerequest) {
+        try {
+            if (roleValidation.isValid(newrolerequest)) {
+                NewRoleRequest newrole = jobService.addNewRole(newrolerequest);
+                return Response.status(HttpStatus.CREATED_201).entity(newrole).build();
+            }else{
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/controller/JobController.java:138]: You can unwrap redundant `else` branch i.e.

```
if (roleValidation.isValid(newrolerequest)) {
                NewRoleRequest newrole = jobService.addNewRole(newrolerequest);
                return Response.status(HttpStatus.CREATED_201).entity(newrole).build();
            }
            return Response.status(HttpStatus.BAD_REQUEST_400).build();
```

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 4]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
...
+            while (rs.next()) {
+             id = rs.getInt(1);
+            }
+
+        } catch (SQLException ex) {
+            ex.printStackTrace();
+        } finally {
+            closeConnection();
+        }
+        if(id <0){
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:294]: You can replace with `Objects.requireNonNull`

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 5]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
...
-}
\ No newline at end of file
+
+    public List<JobRole> populateFamilyLists() throws SQLException {
+
+        List<JobRole> Jobroles = new ArrayList<>();
+
+        try {
+            Connection c = getConnection();
+            String capability = "SELECT capabilityId, capabilityName from jobCapabilities";
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:178]: Object `capability` is never used so can be removed

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 6]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
...
+        } catch (SQLException ex) {
+            ex.printStackTrace();
+        } finally {
+            closeConnection();
+        }
+
+        return Jobroles;
+    }
+
+    public NewRoleRequest Addnewjobrole (NewRoleRequest addedRole) throws SQLException, RoleNotAddedException {
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:267]: Method names should be camel case and start with a lower case

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 7]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
 
         return Competency;
     }
-}
\ No newline at end of file
+
+    public List<JobRole> populateFamilyLists() throws SQLException {
+
+        List<JobRole> Jobroles = new ArrayList<>();
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:174]: Local variable names should start with lower case character

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 8]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
...
+        } finally {
+            closeConnection();
+        }
+
+        return Jobroles;
+    }
+
+    public List<JobRole> populateCapabilityList() throws SQLException {
+
+        List<JobRole> Jobroles = new ArrayList<>();
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:206]: Local variable names should start with lower case character

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 9]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
...
+        } finally {
+            closeConnection();
+        }
+
+        return Jobroles;
+    }
+
+    public List<JobRole> populateBandLevelList() throws SQLException {
+
+        List<JobRole> Jobroles = new ArrayList<>();
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:238]: Local variable names should start with lower case character

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 10]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
...
+
+        try {
+            Connection c = getConnection();
+            String band = "SELECT bandLevelId, BandName from jobBandLevel";
+            PreparedStatement preparedStmt1 = Objects.requireNonNull(c).prepareStatement(band);
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:243]: Investigate using try-with-resources with prepared statement

File: src/main/java/com/kainos/ea/dao/JobDao.java [hunk 11]
```diff
@@ -168,4 +168,132 @@ public List<Competency> getCompetency(int bandID) throws SQLException {
...
+
+        return Jobroles;
+    }
+
+    public List<JobRole> populateBandLevelList() throws SQLException {
+
+        List<JobRole> Jobroles = new ArrayList<>();
+
+        try {
+            Connection c = getConnection();
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/dao/JobDao.java:241]: Variable named `c` does not have a type char or Java.lang.Character. Try using something more descriptive i.e `connection` or `conn`. You can fix rest of the file in same way - thanks

File: src/main/java/com/kainos/ea/model/NewRoleRequest.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
+package com.kainos.ea.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.List;
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/model/NewRoleRequest.java:8]: remove unused import

File: src/main/java/com/kainos/ea/model/NewRoleRequest.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            @JsonProperty("jobFamilyId") int jobFamilyID,
+            @JsonProperty("bandLevelId") int bandLevelID,
+            @JsonProperty("capabilityId") int capabilityID
+    ) {
+        this.setJobName(jobName);
+        this.setJobResponsibility(jobResponsibility);
+        this.setSpecSummary(spec);
+        this.setJobFamilyID(jobFamilyID);
+        this.setBandLevelID(bandLevelID);
+        this.setCapabilityID(capabilityID);
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/model/NewRoleRequest.java:40]: `this` is unnecesary in this context. You can remove

File: src/main/java/com/kainos/ea/model/NewRoleRequest.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.List;
+
+@Getter
+@Setter
+public class NewRoleRequest {
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/model/NewRoleRequest.java:12]: Change class name to `NewRole` is much simpler

File: src/main/java/com/kainos/ea/service/JobService.java [hunk 15]
```diff
@@ -64,5 +66,42 @@ public List<Competency> competency(int bandID) throws DatabaseException, SQLExce
...
+        List<JobRole> capability;
+        capability = jobDao.populateBandLevelList();
+
+        if(capability.size() <1){
+            throw new DatabaseException("Error in view band lists", new Exception());
+        }
+        return capability;
+    }
+
+    public NewRoleRequest addNewRole(NewRoleRequest newjobrole) throws DatabaseException, SQLException, RoleNotAddedException {
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/service/JobService.java:96]: I think we can move this to a new class - `NewRoleService`

File: src/main/java/com/kainos/ea/service/JobService.java [hunk 16]
```diff
@@ -64,5 +66,42 @@ public List<Competency> competency(int bandID) throws DatabaseException, SQLExce
...
+        capability = jobDao.populateFamilyLists();
+        if(capability.size() <1){
+            throw new DatabaseException("Error in view family list", new Exception());
+        }
+        return capability;
+    }
+    public List<JobRole> populateCapabiltyLists() throws DatabaseException, SQLException {
+        List<JobRole> capability;
+        capability = jobDao.populateCapabilityList();
+        if(capability.size() <1){
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/service/JobService.java:80]: You can replace this with `Objects.requireNonNull` or an `assert` statement like so -
                                                 
    `assert capability.size() >= 1 : "Error in view capabilities";`

File: src/main/java/com/kainos/ea/service/JobService.java [hunk 17]
```diff
@@ -64,5 +66,42 @@ public List<Competency> competency(int bandID) throws DatabaseException, SQLExce
...
+            throw new DatabaseException("Error in view capabilities", new Exception());
+        }
+        return capability;
+    }
+
+    public List<JobRole> populateBandLists() throws DatabaseException, SQLException {
+        List<JobRole> capability;
+        capability = jobDao.populateBandLevelList();
+
+        if(capability.size() <1){
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/service/JobService.java:90]: Same as above

File: src/main/java/com/kainos/ea/service/JobService.java [hunk 18]
```diff
@@ -64,5 +66,42 @@ public List<Competency> competency(int bandID) throws DatabaseException, SQLExce
...
+            throw new DatabaseException("Error in view band lists", new Exception());
+        }
+        return capability;
+    }
+
+    public NewRoleRequest addNewRole(NewRoleRequest newjobrole) throws DatabaseException, SQLException, RoleNotAddedException {
+        NewRoleRequest newrole = jobDao.Addnewjobrole(newjobrole);
+
+        if(newrole.getJobName().length() <1){
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/service/JobService.java:99]: You can replace this with `Objects.requireNonNull` or an `assert` statement like so -

` assert newrole.getJobName().length() >= 1 : "Error Adding New Role";`

File: src/main/java/com/kainos/ea/validator/NewJobRoleValidation.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+
+        if(role.getJobName().length() < 5){
+            return false;
+        }
+
+        if(role.getJobResponsibility().length() < 10){
+            return false;
+        }
+
+        if(role.getSpecSummary().length() < 8){
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/validator/NewJobRoleValidation.java:20]: You can probably simplify this last `if` statement like so

`return role.getSpecSummary().length() >= 8;`

File: src/test/java/com/kainos/ea/validator/AddNewRole.java [hunk 20]
```diff
@@ -0,0 +1,109 @@
+package com.kainos.ea.validator;
+
+import com.kainos.ea.model.NewRoleRequest;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AddNewRole {
```
Comment (hellofadude) [inline src/test/java/com/kainos/ea/validator/AddNewRole.java:8]: If this is a test class for the validator you should probably call it `NewRoleValidationTest` to match the corresponding class

File: src/main/java/com/kainos/ea/validator/NewJobRoleValidation.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
+package com.kainos.ea.validator;
+
+import com.kainos.ea.model.NewRoleRequest;
+
+public class NewJobRoleValidation {
```
Comment (hellofadude) [inline src/main/java/com/kainos/ea/validator/NewJobRoleValidation.java:5]: Probably worth changing class name to `NewRoleValidation` - more concise

**Review by hellofadude — APPROVED** [review]

**Review by MihneaBocaKainos — APPROVED** [review]

**Pull request merged**